pkg/pillar/zedcloud/authen.go

```diff
@@ -559,6 +574,82 @@ func verifySignature(log *base.LogObject, certByte []byte, interm *x509.CertPool
 	return nil
 }
 
+func verifyx509TimeStamps(log *base.LogObject, realCert *x509.Certificate) bool {
+	rightNow := time.Now()
+	notBefore := realCert.NotBefore
+	notAfter := realCert.NotAfter
+	certName := realCert.Subject
+	certValid := true
+	// HACK for compilation test DUMMY thresholds
+	type configInfo struct {
+		// only need to define thresholds to
+		//  inform and warn the owners
+		infoThreshold time.Duration
+		warnThreshold time.Duration
+	}
+
+	type certStatus uint8
+	const (
+		allCertsAreValid certStatus = iota
+		aCertIsInfo
+		aCertIsWarn
+		aCertIsINVALID
+	)
+
+	type statusInfo struct {
+		CertStatus certStatus
+	}
+
+	// assume default will be 60 days before for Info
+	//  and 30 days for the Warning
+	var SystemConfig = configInfo{time.Hour * 24 * 60,
+		time.Hour * 24 * 30}
+
+	var CurrCertStatus = statusInfo{allCertsAreValid}
+	// HACK end
+
+	// verify the NotBefore, and NotAfter
+	if rightNow.After(notBefore) {
+		// Cert is active (after NotBefore date)
+		//  process thresholds
+		validDuration := notAfter.Sub(rightNow)
+		switch {
+		case validDuration < 0:
+			log.Errorf("%s: expired %v ago",
+				certName,
+				-validDuration)
+			certValid = false
+			CurrCertStatus.CertStatus = aCertIsINVALID
```

Comment:
I think the function which checks this should also 1) stop accepting new configuration from the controller 2) periodically fetch from /certs to get updated and valid certificate(s), and 3) put the device in maintenance mode until valid certs are received.

We can do 3) as a separate PR now or later; requires definiting a new MaintenanceModeReason value similar to the vaultlocked, and have EVE-OS set that new value and then clear it later. #3 makes sure that things are clearly flagged should be controller send unusable invalid certificate(s).

Reply:
The function posts logs based on the particular threshold.  (Default is 60 days prior a Notice log, escalating to Warning at 30 days -- both deprived from the config, and thus can be administratively set).  Obviously, it posts a Failing log when out of date.

I can refactor this into a function to test the cert against a given date, returning the cert state based on the date specified, and a higher level function that evaluates the cert with the current date/time, and do the logging in a caller to that function.  Thought about that from the start, but opted for what seemed to be the most straight forward approach.

It's about an hour or so to do the refactor.  Let me know if you prefer it that way.

Reply:
There already is a function in the x509 package to to the validation based on a given date.
And I swear I added the name of that function in a comment in this PR a few weeks back; not the first time I find the gh looses comments.

Set CurrentTime with and call https://pkg.go.dev/crypto/x509#Certificate.Verify just like we do elsewhere in EVE-OS code.